python/ray/data/preprocessor.py

```diff
@@ -226,59 +241,83 @@ def _transform(self, dataset: Dataset) -> Dataset:
+                f'"pandas" and "numpy" allowed, but got: {transform_type}'
             )
 
-    def _transform_batch(self, df: "DataBatchType") -> "DataBatchType":
+    def _transform_batch(self, data: "DataBatchType") -> "DataBatchType":
+        # For minimal install to locally import air modules
         import pandas as pd
+        from ray.air.constants import TENSOR_COLUMN_NAME
+        from ray.air.util.data_batch_conversion import (
+            convert_batch_type_to_pandas,
+            _convert_batch_type_to_numpy,
+        )
 
         try:
             import pyarrow
         except ImportError:
             pyarrow = None
 
-        if isinstance(df, pd.DataFrame):
+        if isinstance(data, pd.DataFrame):
             data_format = "pandas"
-        elif pyarrow is not None and isinstance(df, pyarrow.Table):
+        elif pyarrow is not None and isinstance(data, pyarrow.Table):
             data_format = "arrow"
+        elif isinstance(data, (dict, np.ndarray)):
+            data_format = "numpy"
         else:
             raise NotImplementedError(
-                "`transform_batch` is currently only implemented for Pandas DataFrames "
-                f"and PyArrow Tables. Got {type(df)}."
+                "`transform_batch` is currently only implemented for Pandas "
+                "DataFrames, pyarrow Tables, NumPy ndarray and dictionary of "
+                f"ndarray. Got {type(data)}."
             )
 
         transform_type = self._determine_transform_to_use(data_format)
 
         if transform_type == "pandas":
-            if data_format == "pandas":
-                return self._transform_pandas(df)
-            else:
-                return self._transform_pandas(df.to_pandas())
-        elif transform_type == "arrow":
+            return self._transform_pandas(convert_batch_type_to_pandas(data))
+        elif transform_type == "numpy":
+            transformed_numpy_data = self._transform_numpy(
+                _convert_batch_type_to_numpy(data)
+            )
             if data_format == "arrow":
-                return self._transform_arrow(df)
-            else:
-                return self._transform_arrow(pyarrow.Table.from_pandas(df))
+                if len(data.column_names) == 1 and data.column_names != [
+                    TENSOR_COLUMN_NAME
+                ]:
+                    # Single-column table is perserved as a table if not
+                    # explicted named as the tensor column name.
+                    return pyarrow.Table.from_pydict(transformed_numpy_data)
+            elif data_format == "pandas":
+                if len(data.columns) == 1 and list(data.columns) != [
+                    TENSOR_COLUMN_NAME
+                ]:
+                    # Single-column table is perserved as a table if not
+                    # explicted named as the tensor column name.
+                    return pd.DataFrame.from_dict(transformed_numpy_data)
```

Comment:
I'm not really sure why we need this- can it not just be `return self._transform_numpy(_convert_batch_type_to_numpy(data))`?

Out API currently makes no guarantees that the output needs to be the same type as the input: https://github.com/ray-project/ray/blob/master/python/ray/data/preprocessor.py#L158-L161

Do we need to make this guarantee? If not, then we can remove this logic and avoid extra copies.

Reply:
i have no strong preference of either choice, @clarkzinzow had a comment earlier suggesting keep single column. How about we start a thread to discuss this more efficiently rather than adding new things to this PR that extends more iterations?

Reply:
I'm firm on this, if a user defines their data as a single-column table, it needs to be preserved as a single-column table.

Reply:
discussed offline -- at UDF / transform_batch level we still return numpy data type, we can still have single column table in post-processing since numpy data will be converted back to table (arrow/pandas) format